src/mscorlib/src/System/Text/Encoding.cs

```diff
@@ -437,92 +466,34 @@ public static Encoding GetEncoding(int codepage)
-                        // case able to handle more code pages by calling GetEncodingCodePage
-                        case CodePageNoOEM:             // 1
-                        case CodePageNoMac:             // 2
-                        case CodePageNoThread:          // 3
-                        case CodePageNoSymbol:          // 42
-                            // Win32 also allows the following special code page values.  We won't allow them except in the
-                            // CP_ACP case.
-                            // #define CP_ACP                    0           // default to ANSI code page
-                            // #define CP_OEMCP                  1           // default to OEM  code page
-                            // #define CP_MACCP                  2           // default to MAC  code page
-                            // #define CP_THREAD_ACP             3           // current thread's ANSI code page
-                            // #define CP_SYMBOL                 42          // SYMBOL translations
-                            throw new ArgumentException(Environment.GetResourceString(
-                                "Argument_CodepageNotSupported", codepage), "codepage");
-                        // Have to do ASCII and Latin 1 first so they don't get loaded as code pages
-                        case CodePageASCII:             // 20127
-                            result = ASCII;
-                            break;
-                        case ISO_8859_1:                // 28591
-                            result = Latin1;
-                            break;
-                        default:
-                        {
-#if FEATURE_CODEPAGES_FILE
-                            // 1st assume its a code page.
-                            result = GetEncodingCodePage(codepage);
-                            if (result == null)
-                                result = GetEncodingRare(codepage);
-                            break;
-#else
-                            // Is it a valid code page?
-                            if (EncodingTable.GetCodePageDataItem(codepage) == null)
-                            {
-                                throw new NotSupportedException(
-                                    Environment.GetResourceString("NotSupported_NoCodepageData", codepage));
-                            }
-
-                            result = UTF8;
-                            break;
-#endif // FEATURE_CODEPAGES_FILE
-                        }
+                        result = new SBCSCodePageEncoding(codepage);
                     }
-                    encodings.Add(codepage, result);
-                }
+                    else
+                    {
+                        result = GetEncodingCodePage(codepage) ?? GetEncodingRare(codepage);
+                    }
+
+                    Contract.Assert(result != null, "result != null");
 
+                    encodings.Add(key, result);
```

Comment:
@jkotas You may have to comment since I don't work with multithreaded code too much, but I believe the scope of the lock statement can be reduced to just around this `Add`.